File: physics/vector2d.py
class Vector2D:
    def __init__(self, x: float, y: float) -> None:
        self.x: float = x
        self.y: float = y
    
    # When two Vector2D variables are substracted
    def __sub__(self, other: 'Vector2D') -> 'Vector2D':
        return Vector2D(self.x - other.x, self.y - other.y)
    
    # When two Vector2D variables are added
    def __add__(self, other: 'Vector2D') -> 'Vector2D':
        return Vector2D(self.x + other.x, self.y + other.y)
    
    # When two Vector2D variables are multiplied
    def __mul__(self, scalar: float) -> 'Vector2D':
        return Vector2D(self.x * scalar, self.y * scalar)
    
    # When two Vector2D variables are divided
    def __truediv__(self, scalar: float) -> 'Vector2D':
        return Vector2D(self.x / scalar, self.y / scalar)
    
    # When to the power
    def __pow__(self, scalar: float) -> 'Vector2D':
        return Vector2D(self.x ** scalar, self.y ** scalar)
    
    # Reset the position
    def zero(self) -> 'Vector2D':
        self.x = 0.0
        self.y = 0.0
        return self

    # Set the position
    def set(self, x: float = None, y: float = None) -> 'Vector2D': # type: ignore
        if x is not None:
            self.x = x
        if y is not None:
            self.y = y
        return self
    
    # Get the position
    def get(self) -> tuple[float, float]:
        return (self.x, self.y)

    # Dot product
    def dot(self, vec: 'Vector2D') -> float:
        return self.x * vec.x + self.y * vec.y
    
    # Magnitude
    def magnitude(self) -> float:
        return (self.x * self.x + self.y * self.y) ** 0.5
    
    # Copy the vector
    def copy(self) -> 'Vector2D':
        return Vector2D(self.x, self.y)
    
    # Normalize
    def normalize(self) -> 'Vector2D':
        mag: float = self.magnitude()
        return Vector2D(self.x / mag, self.y / mag)
    
    # Scale all values
    def mult(self, other: 'Vector2D') -> 'Vector2D':
        self.x *= other.x
        self.y *= other.y
        return self
    
    # Add to all values
    def add(self, scalar: float) -> 'Vector2D':
        self.x += scalar
        self.y += scalar
        return self
        
    # Substract from all values
    def sub(self, scalar: float) -> 'Vector2D':
        self.x -= scalar
        self.y -= scalar
        return self

    # Divide from all values
    def div(self, other: 'Vector2D') -> 'Vector2D':
        self.x /= other.x
        self.y /= other.y
        return self
    

File: verletsharp/physics/Vector2D.cs
using System;

namespace physics
{
    class Vector2D
    {
        // x and y components of the vector
        public double x;
        public double y;

        // When two vectors are added
        public Vector2D operator +(Vector2D vec)
        {
            return new Vector2D(this.x + vec.x, this.y + vec.y);
        }

        // When two vectors are subtracted
        public Vector2D operator -(Vector2D vec)
        {
            return new Vector2D(this.x - vec.x, this.y - vec.y);
        }

        // When a vector is multiplied by a scalar
        public Vector2D operator Mult(double scalar)
        {
            this.x *= scalar;
            this.y *= scalar;
        }

        // When a vector is divided by a scalar
        public Vector2D Div(double scalar)
        {
            this.x /= scalar;
            this.y /= scalar;
            return this;
        }
        // When a vector is held to the power of a scalar
        public Vector2D Pow(double scalar)
        {
            this.x = Math.Pow(this.x, scalar);
            this.y = Math.Pow(this.y, scalar);
            return this;
        }

        // Add a scalar
        public Vector2D Add(double b)
        {
            this.x += b;
            this.y += b;
            return this;
        }

        // Subtract a scalar
        public Vector2D Sub(double b)
        {
            this.x -= b;
            this.y -= b;
            return this;
        }

        // Multiply by a vector
        public Vector2D MultVector(Vector2D vec)
        {
            this.x *= vec.x;
            this.y *= vec.y;
            return this;
        }

        // Copy the vector
        public Vector2D Copy()
        {
            return new Vector2D(this.x, this.y);
        }

        // Get the vector
        public (double, double) Get()
        {
            return (this.x, this.y);
        }

        // Vector magnitude
        public double Magnitude()
        {
            return Math.Sqrt(this.x * this.x + this.y * this.y);
        }

        // Dot product of two vectors
        public double Dot(Vector2D vec)
        {
            return this.x * vec.x + this.y * vec.y;
        }
    }
}